### ISSUE TITLE
Game resets to attract mode when playing level 21 in demo mode.

### ISSUE THREAD

**Issue by bakerhillpins** [jessaskey/mhavocpe#199]
0.85 PE.  Setup in demo mode and select level 21.  Sit through the tac scan screen. When the VAXX space station appears and zooms towards the player the game will freeze and then reset when it gets to the bottom of the screen.

**Comment by jessaskey** [comment]
This seems to be random and I have seen it happen.... tried to capture it a few times with a trace but good to have it documented here so I don't forget.

**Comment by jessaskey** [comment]
I just tried about 20 times to get this to crash and I was not able to do it now for some reason. Before it was happening fairly regularly, not sure what might have changed that could have 'fixed' this tho... will continue to monitor and leave this open tho.

**Comment by jessaskey** [comment]
As a note on this... there is a flag in the code that I have currently set to NOT reset the VG if it hasn't halted for a long time (generally a math error on writing vector lengths into the VGRAM from the Alpha CPU).... once I release the game, I will put this flag back to the default behavior which will simply reset the VG if it times out instead of crashing the game. Because we are pre-release tho, I have this 'feature' turned off for troubleshooting, so if we can find this bug before release, then I will fix it, but otherwise, it will get swept under the rug with the heavy handed hack/fix. just FYI on this one, I will keep looking tho.

**Comment by bakerhillpins** [comment]
Huh.. this happened every time I tried it on hardware.. I tried 3 times though.

**Comment by jessaskey** [comment]
yeah, it was happening fairly consistently for me on HBMAME for awhile, but for some reason everytime I turn on a trace to try and see what is going on... it doesn't happen... so I will keep trying as there is clearly a random bad RAM variable data value which is causing this... it may be something around the speech again because right when the ship hits the bottom of it's motion is when the speech is triggered. The max/rex collision was causing an issue like this, so I will poke around.

**Comment by jessaskey** [comment]
Okay, I think I actually found this issue... it appears that I had a typo in some random number data loading... details follow..

TLDR: I fixed two lines in code that was causing this crash. 

In the original code, there was one way to get a random number and that was to ask the Gamma for one because the Pokey's which contain the actual random number generators are on the GAMMA processor. It appears that Owen was using this as needed and calling 'jsr random' to return this number. At some point, because the maze sparkoids need 5 random numbers for the directions of sparks, he put this in a loop and stored 5 random numbers in global RAM (rands, rands+1, rands +2, etc)  via the main game loop. In much of MY code, I was calling 'jsr random' to get random numbers and masking them to determine random actions on 1/2,1/4,1/8, etc probabilities...  at some point I was having issues with the GAMMA getting clobbered with requests, so I realized that I should just query one of the random numbers already filled from the main loop. So, I did a search for 'random' and replaced it with 'rands'.... apparently in TWO location, I left a 'jsr rands' instead of 'lda rands', which are substantially different results... the 'jsr' to the RAM location of RANDS will give fairly bad results generally and cause a crash.

**Comment by bakerhillpins** [comment]
This is fixed in 0.86 PE.
